### PR TITLE
[iwyu_tool] Default to -j=0 and infer the core count in that case

### PR DESCRIPTION
**Pull request by pvutov**
Make the behavior of the -j argument more similar to other tools which offer it. Instead of hanging forever when "zero jobs" are requested, attempt to automatically pick a job count suitable for the current system. Default to this behavior if the user has not explicitly specified some other job count.

### CODE REVIEW

**Comment by kimgr** [comment]
Nice, thanks! I didn't know about the -j0 convention.

**Comment by pvutov** [comment]
On second thought, I may have overstated the "convention".

- GNU make offers `make -j`. If no value is provided, the job count is automatically determined. If 0 is provided, they actually complain about wanting a positive integer. If `-j` is not provided at all, runs single-threaded.

- ninja runs multithreaded with or without `-j`, and also supports `-j=0`

- `run-clang-tidy` accepts both `-j` and `-j=0`.

- `cppcheck` does not accept `-j` and hangs on `-j=0` (same as current iwyu).

tldr I will update the patch tonight to also support `-j` with no value. I assume it's fine to run multithreaded if the user has not provided `-j` at all?

**Comment by kimgr** [comment]
National holiday yesterday, so I didn't get a chance to look into this.

> I assume it's fine to run multithreaded if the user has not provided -j at all?

I would actually prefer to be single threaded unless -j is added. I'll take a closer look at the patch later today.

File: iwyu_tool_test.py [hunk 1]
```diff
@@ -363,6 +363,28 @@ def test_extra_args_with_sep(self):
...
+        self.assertNotEqual(None, self.main.call_args['jobs'])
+
+    def test_jobs_novalue(self):
+        """ It is allowed to use -j without a value """
+        argv = ['iwyu_tool.py', '-j', '-p', '.']
+        iwyu_tool._bootstrap(argv)
+        self.assertNotEqual(0, self.main.call_args['jobs'])
+        self.assertNotEqual(None, self.main.call_args['jobs'])
+
+
```
Comment (kimgr) [inline iwyu_tool_test.py:387]: I'd like to see a testcase here for a command-line without `-j` -- I think that will currently evaluate to `jobs == 0`.

File: iwyu_tool.py [hunk 2]
```diff
@@ -496,8 +496,11 @@ def custom_help():
     parser.add_argument('-o', '--output-format', type=str,
                         choices=FORMATTERS.keys(), default=DEFAULT_FORMAT,
                         help='Output format (default: %s)' % DEFAULT_FORMAT)
-    parser.add_argument('-j', '--jobs', type=int, default=1,
-                        help='Number of concurrent subprocesses')
+    parser.add_argument('-j', '--jobs', type=int, default=0,
```
Comment (kimgr) [inline iwyu_tool.py:499]: It's a bit tricky, but I'd like to see the absence of `-j` evaluate to jobs == 1, and `-j` without a value evaluate to jobs == 0 (which is further refined to `os.cpu_count()`).

Is `nargs='?'` required to make the integer value optional? I'm not sure what effect `const=0` has here.

I can't think of an argparse flag configuration to allow `() => 1`, `-j => 0`, `-j4 => 4`, so I wonder if it's easier to just remove all the argparse settings and check `args.jobs` using `hasattr`/`getattr`

File: iwyu_tool_test.py [hunk 3]
```diff
@@ -363,6 +363,28 @@ def test_extra_args_with_sep(self):
...
                          self.main.call_args['extra_args'])
 
+    def test_jobs_normal(self):
+        """ Nonzero values for -j are forwarded as they are. """
+        argv = ['iwyu_tool.py', '-p', '.', '-j', '4']
+        iwyu_tool._bootstrap(argv)
+        self.assertEqual(4, self.main.call_args['jobs'])
+
+
+    def test_jobs_auto(self):
```
Comment (kimgr) [inline iwyu_tool_test.py:373]: `test_jobs_zero` seems more apt for this one

File: iwyu_tool.py (hunk 2 above)
Comment (pvutov) [reply iwyu_tool.py:499]: `const=0` determines the value if the argument is used as a flag (`-j` case, as opposed to `-jX`). To get what you want, I just need to change to `default=1, const=0`.

Comment (kimgr) [reply iwyu_tool.py:499]: Ah, nice. I learn something new about argparse every time I use it :upside_down_face:

File: iwyu_tool_test.py [hunk 4]
```diff
@@ -363,6 +363,34 @@ def test_extra_args_with_sep(self):
...
 
+    def test_jobs_normal(self):
+        """ Nonzero values for -j are forwarded as they are. """
+        argv = ['iwyu_tool.py', '-p', '.', '-j', '4']
+        iwyu_tool._bootstrap(argv)
+        self.assertEqual(4, self.main.call_args['jobs'])
+
+
+    def test_jobs_zero(self):
+        """ -j=0 will not hang the program. """
```
Comment (kimgr) [inline iwyu_tool_test.py:374]: Hmm, I don't think this testcase demonstrates `-j=0` doesn't hang, just that zero translates to some arbitrary non-zero?

File: iwyu_tool_test.py [hunk 5]
```diff
@@ -363,6 +363,34 @@ def test_extra_args_with_sep(self):
...
+        argv = ['iwyu_tool.py', '-p', '.', '-j', '0']
+        iwyu_tool._bootstrap(argv)
+        self.assertNotEqual(0, self.main.call_args['jobs'])
+        self.assertNotEqual(None, self.main.call_args['jobs'])
```
Comment (kimgr) [inline iwyu_tool_test.py:378]: Prefer `assertIsNotNone` for None-checking: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIsNotNone

File: iwyu_tool_test.py [hunk 6]
```diff
@@ -363,6 +363,34 @@ def test_extra_args_with_sep(self):
...
+
+    def test_jobs_novalue(self):
+        """ It is allowed to use -j without a value """
+        argv = ['iwyu_tool.py', '-j', '-p', '.']
+        iwyu_tool._bootstrap(argv)
+        self.assertNotEqual(None, self.main.call_args['jobs'])
```
Comment (kimgr) [inline iwyu_tool_test.py:384]: `assertIsNotNone`

File: iwyu_tool_test.py [hunk 7]
```diff
@@ -363,6 +363,34 @@ def test_extra_args_with_sep(self):
...
+        argv = ['iwyu_tool.py', '-p', '.', '-j', '4']
+        iwyu_tool._bootstrap(argv)
+        self.assertEqual(4, self.main.call_args['jobs'])
+
+
+    def test_jobs_zero(self):
+        """ -j=0 will not hang the program. """
+        argv = ['iwyu_tool.py', '-p', '.', '-j', '0']
+        iwyu_tool._bootstrap(argv)
+        self.assertNotEqual(0, self.main.call_args['jobs'])
```
Comment (kimgr) [inline iwyu_tool_test.py:377]: Should this be `assertLess` too, like below?

File: iwyu_tool_test.py (hunk 4 above)
Comment (pvutov) [reply iwyu_tool_test.py:374]: I guess I just change the comment?

File: iwyu_tool.py [hunk 8]
```diff
@@ -497,7 +497,10 @@ def custom_help():
                         choices=FORMATTERS.keys(), default=DEFAULT_FORMAT,
                         help='Output format (default: %s)' % DEFAULT_FORMAT)
     parser.add_argument('-j', '--jobs', type=int, default=1,
-                        help='Number of concurrent subprocesses')
+                        nargs='?', const=0,
+                        help='Number of concurrent subprocesses. '
```
Comment (kimgr) [inline iwyu_tool.py:501]: This fits on two lines with 80-column width:
```
                        help='Number of concurrent subprocesses. If zero, will '
                             'try to match the logical cores of the system.')

```

File: iwyu_tool_test.py (hunk 4 above)
Comment (kimgr) [reply iwyu_tool_test.py:374]: Yeah, that makes sense. I was trying to come up with a way of testing not hanging, but these tests are heavily mocked, so I don't think that's straightforward.

File: iwyu_tool.py (hunk 8 above)
Comment (kimgr) [reply iwyu_tool.py]: Oops, lost the leading parenthesis, let's do it as:
```
                        help=('Number of concurrent subprocesses. If zero, '
                              'will try to match the logical cores of the '
                              'system.')
```
to help editors that understand ()-wrapped multiline strings.

Comment (kimgr) [reply iwyu_tool.py]: See the `source` argument for conventional wrapping. The others seem to have snuck through review before.

File: iwyu_tool.py [hunk 9]
```diff
@@ -497,9 +497,13 @@ def custom_help():
                         choices=FORMATTERS.keys(), default=DEFAULT_FORMAT,
                         help='Output format (default: %s)' % DEFAULT_FORMAT)
     parser.add_argument('-j', '--jobs', type=int, default=1,
-                        help='Number of concurrent subprocesses')
+                        nargs='?', const=0,
+                        help=('Number of concurrent subprocesses. If zero, '
+                              'will try to match the logical cores of the '
+                              'system.'))
     parser.add_argument('-l', '--load', type=float, default=0,
-                        help='Do not start new jobs if the 1min load average is greater than the provided value')
```
Comment (kimgr) [inline iwyu_tool.py:502]: Ah, sorry, please revert this change, it has nothing to do with the jobs patch.

I'll push a refactor to clean these up after.

Comment (kimgr) [reply iwyu_tool.py:502]: Oh, I just saw this was in a separate commit. That's fine, let's leave it in.

**Comment by kimgr** [comment]
Thanks, and sorry about all the back and forth!
